delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/Permissions.scala

```diff
@@ -71,3 +79,118 @@ trait Permissions {
     */
   def delete(rev: Long): IO[PermissionsRejection, PermissionsResource]
 }
+
+object Permissions {
+
+  private def withPermissions(
+      state: PermissionsState,
+      minimum: Set[Permission]
+  )(permissions: Set[Permission], instant: Instant, subject: Subject): PermissionsState = {
+    state match {
+      case Initial          =>
+        Current(
+          rev = 1L,
+          permissions = permissions ++ minimum,
+          createdAt = instant,
+          createdBy = subject,
+          updatedAt = instant,
+          updatedBy = subject
+        )
+      case current: Current =>
+        current.copy(
+          rev = state.rev + 1,
+          permissions = permissions ++ minimum,
+          updatedAt = instant,
+          updatedBy = subject
+        )
+    }
+  }
+
+  private[delta] def next(
+      minimum: Set[Permission]
+  )(state: PermissionsState, event: PermissionsEvent): PermissionsState = {
+    val apply                                                  = withPermissions(state, minimum) _
+    def appended(e: PermissionsAppended): PermissionsState     =
+      state match {
+        case Initial if e.rev == 1L           => apply(e.permissions, e.instant, e.subject)
+        case s: Current if s.rev + 1 == e.rev => apply(s.permissions ++ e.permissions, e.instant, e.subject)
```

Comment:
We will find this condition on revision for every state/event, we may find a way to abstract it ?

Reply:
The state machine is quite self containing and explicit. I don't think abstracting in this situation is helping to understand the code

Reply:
In your PR, there is no check on revision when applying the event and it is true that an incorrect revision must be caught there. So maybe, we just don't need them at all at this point ?

Reply:
You are right. The check on rev is done on the evaluate function. The next function should not do that, it is not its responsibility.